docs(step_15): create new step about accessibility

felixzapata opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Docs update to add a new step about accessibility

**What is the current behavior? (You can also link to an open issue here)**
The current tutorial is not accessible. With this new step you can learn [how to fix the issues](https://github.com/angular/angular-phonecat/pull/341).

**What is the new behavior (if this is a feature change)?**
Introduce the accessibility into the tutorial.

**Does this PR introduce a breaking change?**
No, it does'nt. Only affects to the tutorial.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:
My opinion is this step should be a process inside of the tutorial. Not a separate step. This way the user can understand that is optional but it should be mandatory.

felixzapata committed:
docs(step_15): create new step about accessibility

gkalpak commented:
fyi @marcysutton

gkalpak commented on docs/content/tutorial/step_15.ngdoc:
```diff
@@ -0,0 +1,327 @@
+<div doc-tutorial-reset="15"></div>
+
+## Why accessibility matters
+
```
Could you add a paragraph on what is "accessibility" and "aria" and how it affects people. Just a brief summary and then list the links for more info.

felixzapata replied:
I've added some definitions.

gkalpak commented on docs/content/tutorial/step_15.ngdoc:
````diff
@@ -0,0 +1,327 @@
+```html
+  <div>
+    <label for="search">Search:</label> 
+    <input id="search" ng-model="$ctrl.query" ng-model-options="{ debounce: 900 }">
````
Why this large debounce delay? Is it necessary? Does the delay have to be that long?

felixzapata replied:
My first attempt was not to use the debounce, but without it, there problems to inform the user about the number of results. The information that the user obtains is not correct. It is neccesary an small "gap" to get the results, and update the aria region.

I can try to reduce this number to the mininum required.

felixzapata replied:
I've reduced the debounce to 300.

gkalpak commented on docs/content/tutorial/step_15.ngdoc:
```diff
@@ -0,0 +1,327 @@
+
+## How to use the `ngAria` module
+
+The `ngAria` module provides support for common ARIA attributes that convey state or semantic information about the application for users of assistive technologies, such as screen readers.
```
We should also have a paragragh on how our specific app will be benefited just by including `ngAria` (to give more emphasis on what `ngAria` does under the hood for us).

gkalpak commented on docs/content/tutorial/step_15.ngdoc:
```diff
@@ -0,0 +1,327 @@
+
+The current detail of a phone prevents a user to use a keyboard to navigate for over each thumbnail image. It is very important due to accessibility that all the actions and information are available via keyboard too. For example, lightboxes, carousels, slides, etc.
+
+_Tip_: Try to navigate over the application using only the _TAB_ key. You will discover that some actions can't be accessed nor the focus on some elements. 
```
So, how are we solving this?

felixzapata replied:
I've added how to fix this problem (adding a button element). For this reason, and after check again the project, the `ngAria` module is not necessary (I added it, to activate the `ng-click` inside the `img` element).

gkalpak commented:
I left a few comments. Overall looks good :+1:

A couple more necessary changes:
- Limit lines to 100 chars max.
- Don't mention anything protractor-related until the "Testing" section (except for the step summary at the top).
- I would rather not introduce new features at this point. Custom `directives`, `$scope.$watch` etc, are not explained before. Are they necessary?

felixzapata commented:
For this example, I think is better the directives rather to use a component. These directives are needed to intercept the repeater and return the values to the aria regions. Maybe this can do it in another way.

Another option would be to remove the section "How to use the directives" and add a link to the [Directives webpage](http://localhost:8000/build/docs/guide/directive).

felixzapata committed:
docs(step_15.ngdoc): update the step 15

add definitions about accessibility and WAI-ARIA
ngAria module now is not neccesary, so the section has been moved to explain how to use it if you need it
explain how to fix the problem with the keyboard in the detail page.

felixzapata commented:
@gkalpak I've made some updates. I will ping you, when I complete all of them (limit of lines, explanation of protractor accessibility plugin).

marcysutton commented on docs/content/tutorial/step_15.ngdoc:
```diff
@@ -0,0 +1,337 @@
+* [Accessibility in AngularJS and Beyond](http://marcysutton.com/slides/fluent2015/)
+* [Introduction to Web Accessibility](https://www.w3.org/WAI/intro/accessibility.php)
+* [Apps For All: Coding Accessible Web Applications](https://shop.smashingmagazine.com/products/apps-for-all)
+
```
Here's one more article that was written in response to Angular a11y: https://www.smashingmagazine.com/2015/05/client-rendered-accessibility/

marcysutton commented on docs/content/tutorial/step_15.ngdoc:
```diff
@@ -0,0 +1,337 @@
+
+WAI-ARIA (_the Accessible Rich Internet Applications Suite_) gives you the ability to reclassify and otherwise augment the perceived meaning (or semantics) of your HTML.
+
+ Currently certain functionality used in Web sites is not available to some users with disabilities, especially people who rely on screen readers and people who cannot use a mouse. WAI-ARIA addresses these accessibility challenges, for example, by defining new ways for functionality to be provided to assistive technology. With WAI-ARIA, developers can make advanced Web applications accessible and ...
```
I would mention that ARIA is a standard set of attributes used to fill accessibility support gaps in custom HTML elements and SVG.

marcysutton commented on docs/content/tutorial/step_15.ngdoc:
````diff
@@ -0,0 +1,337 @@
+  
+```
+
+The new element is a _region live_. It means that every time this region is updated, it will inform the user about its content.
````
*live region

marcysutton commented on docs/content/tutorial/step_15.ngdoc:
```diff
@@ -0,0 +1,337 @@
+
+To fix this issue, we add a `button` element, wrapping the image to allow the user to access the content via keyboard.
+
+Why we are using a `button` element instead of an `a` element? It is more correct to use buttons elements to call actions like show content, hidden content or similar.
```
Because it does not navigate to a new page or resource.

felixzapata committed:
docs(step_15): update document following marcysutton comments

felixzapata commented:
@gkalpak the limit of 100 characters is only for the code examples or all the document?

felixzapata committed:
docs(step_15): add more information about the protractor plugin

gkalpak commented:
@felixzapata, the limit is for the whole document (where ever possible, e.g. if a URL is too long it's OK).
Is this ready for another review?

felixzapata commented:
@gkalpak I have to review about the lines and I have to think what to do with the directives. Have you seen my previous comment about that? I mean, remove the section "How to use the directives" and add a link to the Directives webpage, instead of copy the code of the directives.

felixzapata committed:
docs(step_15): update document

felixzapata committed:
docs(step_15): update the doc

remove information about directives. Now the tutorial links to the guide.
remove information about how to install ngAria because is not used in this step.

felixzapata commented:
@gkalpak is ready for another review:

- I've removed the information about directives. Now the tutorial has a link to the directives guide.
- I've removed the information about how to instal the ngAria module becuase, finally, is not used in this step.

When the step is approved, I will squash the commits.

felixzapata commented:
@gkalpak it is ready for another review.

felixzapata commented:
@gkalpak hi, any news about this pull request?

felixzapata commented:
I signed it!

gkalpak commented:
Sorry it is taking so long, @felixzapata. It is still on my todo list. I'll get to it after 1.5.9 and 1.6.0-rc.0 are released.

felixzapata commented:
ok @gkalpak
